pkg/vault/test/foundry/BufferVaultPrimitive.t.sol

```diff
@@ -508,6 +509,134 @@ contract BufferVaultPrimitiveTest is BaseVaultTest {
+        uint256 expectedWrappedOut = proportionalWithdrawPct.mulDown(bufferWrappedBalance);
+        // Will get 1333.333/3333.333 = 40% of value:
+        // [0.4 * 3000, 0.4 * 1000] = [1200 underlying, 400 wrapped] - worth 2000 = amount in
+        vm.prank(lp);
+        (uint256 removedUnderlying, uint256 removedWrapped) = vault.removeLiquidityFromBuffer(waDAI, secondAddLpShares);
+        assertApproxEqAbs(removedUnderlying, expectedUnderlyingOut, 1e6, "Wrong underlying amount removed");
+        assertApproxEqAbs(removedWrapped, expectedWrappedOut, 1e6, "Wrong wrapped amount removed");
+
+        uint256 totalUnderlyingValue = removedUnderlying + rate.mulDown(removedWrapped);
+        assertLe(totalUnderlyingValue, secondAddUnderlying, "Value removed > value added");
+        assertApproxEqAbs(totalUnderlyingValue, secondAddUnderlying, 3, "Value removed !~ value added");
+    }
+
+    // Trying to increase the coverage by splitting into two rate regimes, and limiting the range.
+    function testAddLiquidityToBufferWithIncreasedRate_Fuzz(
+        uint128 firstDepositUnderlying,
+        uint128 secondDepositUnderlying,
+        uint128 wrappedAmount,
+        uint64 rate
+    ) public {
+        _addLiquidityToBufferWithRate(
+            bound(firstDepositUnderlying, 0, _userAmount),
+            bound(secondDepositUnderlying, 0, _userAmount),
+            bound(wrappedAmount, 0, _userAmount),
+            bound(rate, 1e18, 10_000e18)
+        );
+    }
+
+    function testAddLiquidityToBufferWithDecreasedRate_Fuzz(
+        uint128 firstDepositUnderlying,
+        uint128 secondDepositUnderlying,
+        uint128 wrappedAmount,
+        uint64 rate
+    ) public {
+        _addLiquidityToBufferWithRate(
+            bound(firstDepositUnderlying, 0, _userAmount),
+            bound(secondDepositUnderlying, 0, _userAmount),
+            bound(wrappedAmount, 0, _userAmount),
+            bound(rate, 0.0001e18, 1e18)
+        );
+    }
+
+    function _addLiquidityToBufferWithRate(
+        uint256 firstDepositUnderlying,
+        uint256 secondDepositUnderlying,
+        uint256 wrappedAmount,
+        uint256 rate
+    ) internal {
+        // Ensure we're adding more than the minimum, or it will revert.
+        vm.assume(firstDepositUnderlying + wrappedAmount >= MINIMUM_TOTAL_SUPPLY);
+
+        vm.prank(lp);
+        uint256 firstAddLpShares = router.initializeBuffer(waDAI, firstDepositUnderlying, wrappedAmount);
+        assertEq(
+            firstAddLpShares,
+            firstDepositUnderlying + wrappedAmount - MINIMUM_TOTAL_SUPPLY,
+            "Wrong first lpShares added"
+        );
+
+        // Change the rate after initialization.
+        waDAI.mockRate(rate);
+
+        // Predict the amount of shares to receive
+        (uint256 bufferUnderlyingBalance, uint256 bufferWrappedBalance) = vault.getBufferBalance(waDAI);
+        uint256 currentInvariant = bufferUnderlyingBalance + bufferWrappedBalance.mulDown(rate);
```

Comment:
```suggestion
        uint256 currentInvariant = bufferUnderlyingBalance + waDAI.convertToAssets(bufferWrappedBalance);
```

Multiplying by rate is different than convertToAssets. Multiply by the rate is the same as doing `bufferWrappedBalance.mulDown(underlyingBalanceOfWrapper).divDown(wrapperTotalSupply)`, which introduces rounding issues since mulDown divides by 1e18. 

ConvertToAssets is `bufferWrappedBalance * underlyingBalanceOfWrapper / wrapperTotalSupply`, which is a more precise operation.